src/omnipool/math.rs

```diff
@@ -242,13 +243,36 @@ pub fn calculate_add_liquidity_state_changes(
     })
 }
 
+/// Calculate withdrawal fee given current spot price and oracle price.
+pub fn calculate_withdrawal_fee(
+    spot_price: FixedU128,
+    oracle_price: FixedU128,
+    min_withdrawal_fee: Permill,
+) -> Option<FixedU128> {
+    let price_diff = if oracle_price <= spot_price {
+        spot_price.saturating_sub(oracle_price)
+    } else {
+        oracle_price.saturating_sub(spot_price)
+    };
+
+    // just a defensive check to ensure that min withdrawal fee is really <= 1.
+    // Reason being that clamp panics if min value > max value.
+    let min_fee: FixedU128 = min_withdrawal_fee.into();
```

Comment:
permill can be more than 1 ?

Reply:
no it cannot. yes, it is impossible for this scenario to happen atm. but i figured, rather safe than sorry, hence the defensive comment.

Reply:
also what is user supposed to do if they get None ? there is no fee? infinite fee, this case has to be defined

Reply:
you are clamping result to the 1 anyway, so i dont really see benefit of introduction of option here

Reply:
> you are clamping result to the 1 anyway, so i dont really see benefit of introduction of option here, if you want to have error state, just return error, not None. we should benefit from rust here, not obfuscate because of it

yes, agree. initial implementation i had it without option as i could always return a fee. 

the check to ensure that the min fee is correct changed that.

but thinking of that now - it is not needed as it is not possible to happen. We can ensure it anyway just by using debug asserts only. 

will refactor. i like it more that way

Reply:
> even if there would be possible to fail calculation based on the input, i would prefer to return result with error instead of option with none

another erason why there is option is the checked div. but this can be avoided by simply adding a check if oracle price is zero. 

if the price is zero - remove liquidity fails ( there is explicit check in the pallet itself )

Regarding the result vs option - if any calculation fails, i prefer option. if there is a case where a condition fails ( like here ) - then result if preferable, yes.

Reply:
refactored. let me know what you think

Reply:
not sure when could be the debug assert triggered as we run even tests with the `--release` flag most of the time